### ISSUE TITLE
test performance, estimate training time

### ISSUE THREAD

**Issue by APJansen** [UNSAT3D/unsat#7]
To apply for a budget on snellius, we need a rough estimate on how many hours we need.

We should first make sure that the data loading is happening efficiently, and then run at least one epoch with a decent sized U-Net, ideally on snellius GPUs (I think I can use another project's budget for a small test), multiply that by an estimate of how many epochs we need to train for, and finally multiply that by 100 to allow for experimentation with various models. If that results in a reasonable number we'll go with that.

**Comment by PabRod** [comment]
We can put this on hold. Our collaborators already managed the space for us.

**Comment by APJansen** [comment]
Currently it's super slow, I'm looking into this now.

Training for 1 epoch on a training set of 540 samples (full (650,650,1) slices), using a very simple model with two hidden layers of 10 units each, takes exactly 1 minute.
I compared this to a simple test using Keras, with random data of the same size, and the same model, where it takes only 10 seconds
It could be multiple things:
- data loading being slow
I tried to do this right from the start, and I compared to a test training set which just creates the data in memory as torch tensors, and it was exactly as slow.
- Not properly making use of the GPU
The GPU is detected, and shown as being in use. Perhaps a test on snellius would be good.
- Overhead
I tried removing any metrics, any weights and biases logging, seems like nothing non-essential is being done, but it doesn't make any difference. Perhaps it's overhead from lightning..?

I'll check on snellius, and if that doesn't solve it, before spending more time on this, I'll check how it scales with model size. If it doesn't, perhaps it's not the most important issue right now.

**Comment by APJansen** [comment]
The problem does remain upon increasing model size, so it's not a small constant overhead.

However, it's a lot better on snellius: found 1:20 locally, 13s using Keras locally, but 18s on Snellius, so that seems close enough for now.
Although something to investigate further is that locally, the operation `aten::copy_` takes up 90% of the cpu time, and on snellius `cudaMemcpyAsync` takes up 86%. The main difference is of course the GPU type, could be that this is not an issue on snellius because the CPU operations happen in parallel, maybe this goes wrong locally or something.
Also the profile on my MAC doesn't show GPU usage, on Snellius it does.
